### PR TITLE
Datasets memcap/v2

### PR DESCRIPTION
**Pull request by inashivb**
Known fixes to be done:
- Unnecessary Debug statements
- TODO corresponding to `GetDefaultMemcap`
- Commit history and messages
- Docs updation

### CODE REVIEW

File: src/datasets.c [hunk 1]
```diff
@@ -593,11 +609,45 @@ void DatasetPostReloadCleanup(void)
     SCMutexUnlock(&sets_lock);
 }
 
+static void GetDefaultMemcap(uint64_t *memcap, uint32_t *hashsize) // TODO Do something about this name! OR Save default memcap in dataset struct?
```
Comment (inashivb) [inline src/datasets.c:612]: @victorjulien could you please check if this is a decent way of retrieving defaults from yaml? This shall be called per rule and once during Init.

Comment (victorjulien) [reply src/datasets.c:612]: might be able to simplify this to something like:

```c
const char *str = NULL;
if (ConfGetValue("datasets.defaults.memcap", &str) == 1) {
    if (ParseSizeStringU64(str, memcap) < 0) {
        FatalError(SC_ERR_FATAL, "memcap value cannot be deduced: %s", str);
    }
}
```
I think the yaml part is a map, not a list, so we can avoid the loop.

Comment (inashivb) [reply src/datasets.c:612]: OK. Thank you!

File: src/util-thash.c [hunk 2]
```diff
@@ -320,6 +329,14 @@ THashTableContext* THashInit(const char *cnf_prefix, size_t data_size,
     return ctx;
 }
 
+/* \brief Set memcap to current memuse
+ * */
+void THashConsolidateMemcap(THashTableContext *ctx)
+{
+    ctx->config.memcap = SC_ATOMIC_GET(ctx->memuse);
```
Comment (victorjulien) [inline src/util-thash.c:336]: wondering if we should update this to make the memcap something like `MAX(SC_ATOMIC_GET(ctx->memuse), default_memcap)`. If the configs memcap is higher than the memuse of the loaded set, use the memcap setting instead.

Comment (inashivb) [reply src/util-thash.c:336]: Makes sense.

File: src/datasets.c [hunk 3]
```diff
@@ -593,11 +609,45 @@ void DatasetPostReloadCleanup(void)
...
+{
+    ConfNode *datasets = ConfGetNode("datasets");
+    if (datasets != NULL) {
```
Comment (jlucovsky) [inline src/datasets.c:615]: nit/style: If `datasets == NULL`, return. This will remove one level of indentation

File: src/detect-dataset.c [hunk 4]
```diff
@@ -181,7 +184,7 @@ static int DetectDatasetParse(const char *str,
                         "'load' can only appear once");
                     return -1;
                 }
-                SCLogDebug("load %s", val);
+                SCLogDebug("load %s size: %lu", val, load_size);
```
Comment (victorjulien) [inline src/detect-dataset.c:187]:
```
Invalid printf with size_t (not casted to uintmax_t) at /builds/inliniac/suricata-ci/suricata/src/detect-dataset.c:187
Makefile:545: recipe for target 'check' failed
```
To print a `sizet_t` use `SCLogDebug("load %s size: %"PRIuMAX, val, (uintmax_t)load_size);`

**Comment by inashivb** [comment]
Replaced with #5401

**Pull request closed without merging**